### DISCUSSION TITLE
How Stdin/Stdout used by fzf?

### DISCUSSION THREAD

**Discussion by antonmedv** [junegunn/fzf#2686] [General]
Hi,

Question: how fzf manage to write the result to stdout but still open interactive TUI in alt screen?

How this is done? `ls | fzf | cat`

Thanks.

**Comment by junegunn** [comment]
The example is misleading, you should remove `$(...)` and change `echo` to `cat` to make your point.

```sh
ls | fzf | cat
```

Having said that, to answer the original question, fzf prints its UI to STDERR instead of STDOUT.

**Reply by antonmedv** [reply]
Ohh. Nice))

**Reply by antonmedv** [reply]
I didn’t thought about it. Thanks for answering.

**Reply by leiless** [reply]
Hi, @junegunn. Does the `fzf` always print the tui in STDERR, or in some specific conditions?

**Reply by leiless** [reply]
Update: After `grep`-ing the `src/tui` folder, I found that the STDERR seems solely used in displaying the tui?

**Reply by junegunn** [reply]
Always.

**Comment by manyids2** [comment]
Hi,
To follow up, how does fzf read from STDIN for user input while a long running command is piped to it?
New to TUI, so assuming the pipe writes to STDIN of fzf, it should be blocked till STDIN is drained?
For example 
`ls -R / | fzf `

**Reply by junegunn** [reply]
fzf reads `/dev/tty` directly.

**Reply by manyids2** [reply]
Oh ok! Thank you for replying!

Out of context, but you're a genius! Thank you so much for fzf!
